### PR TITLE
Disable no-underscore-dangle rule

### PR DESCRIPTION
**Pull request by rawagner**
The rule does not allow underscore in plugin components but is widely used in console. I'd like to disable it to allow plugin components use the same convention as rest of console's code.

See https://github.com/openshift/console/pull/1803#discussion_r297365651

@vojtechszocs @spadgett @christianvogt any objections ?

### CODE REVIEW

**Comment by christianvogt** [comment]
I'm in favor of this rule to enforce use of typescript private vs underscore naming convention. I'm not sure if there's a way to allow ending a name with underscore while preventing the prefix.

I think a better convention would be to name classes according to their hoc enhancement rather than adding an underscore suffix to the base component.

**Comment by spadgett** [comment]
We might be able to use `enforceInMethodNames`

https://eslint.org/docs/rules/no-underscore-dangle#enforceinmethodnames

> I think a better convention would be to name classes according to their hoc enhancement rather than adding an underscore suffix to the base component.

Agree. It would be good to come up with a consistent naming convention for connected components.

**Comment by christianvogt** [comment]
If we adopt the convention of exposing a single default component per file, then this problem goes away as the hoc enhanced component is default exported inline and didn't require a name.

**Comment by vojtechszocs** [comment]
:+1: on using TypeScript `private` modifier in favor of underscore naming convention.

However, this doesn't apply to React (function or class) component declarations, where a trailing underscore was used to denote that the given component is internal / inner / part of another component.

> I'm not sure if there's a way to allow ending a name with underscore while preventing the prefix.

Sadly, the [no-underscore-dangle](https://eslint.org/docs/rules/no-underscore-dangle) rule doesn't seem to support configuration that only allows single trailing underscore for function or class declarations.

> I think a better convention would be to name classes according to their hoc enhancement rather than adding an underscore suffix to the base component.

This reminds me of some projects using e.g. `ConnectedFoo` for `Foo` component. But as @rawagner said, we already have the trailing underscore applied consistently within the `public` directory.

> We might be able to use `enforceInMethodNames`

Yes, but that only covers object property names, not declared variable names themselves.

> If we adopt the convention of exposing a single default component per file, then this problem goes away as the hoc enhanced component is default exported inline and didn't require a name.

I agree with that. Code under `packages` should follow single default component export per file convention.

If some package isn't ready to follow that convention, it can add e.g. `packages/foo-pkg/.eslintrc` file that extends `packages/.eslintrc.js` and suppresses `no-underscore-dangle` until its code is updated.

If we decide to follow some convention, we should apply it consistently to all code under `packages`, keeping in mind that `public` code will be moved to `packages` eventually.

**Comment by vojtechszocs** [comment]
TL;DR version

> If some package isn't ready to follow that convention, it can add e.g. `packages/foo-pkg/.eslintrc` file that extends `packages/.eslintrc.js` and suppresses no-underscore-dangle until its code is updated.

**Comment by christianvogt** [comment]
@vojtechszocs well put. Override for the particular rules only within the particular package. This should be seen as a temporary fix and teams should then work towards adopting the primary rule set.

**Comment by rawagner** [comment]
Ok, thank you all for discussion :)

Lets keep the rule enabled and follow convention of exposing a single default component per file. If some plugin isn't ready to do that they could disable the rule in their `.eslintrc.js`

Closing.
